File: src/main/scala/ru/tinkoff/gatling/amqp/request/RequestReplyDslBuilderExchange.scala
package ru.tinkoff.gatling.amqp.request

import io.gatling.core.Predef._
import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression

case class RequestReplyDslBuilderExchange(requestName: Expression[String], configuration: GatlingConfiguration) {
  def directExchange(name: Expression[String], routingKey: Expression[String]): RequestReplyDslBuilderMessage =
    destination(AmqpDirectExchange(name, routingKey))

  def topicExchange(name: Expression[String], routingKey: Expression[String]): RequestReplyDslBuilderMessage =
    destination(AmqpTopicExchange(name, routingKey))

  def queueExchange(name: Expression[String]): RequestReplyDslBuilderMessage = destination(AmqpQueueExchange(name))

  def destination(dest: AmqpExchange) = RequestReplyDslBuilderMessage(
    requestName,
    dest,
    AmqpQueueExchange(""),
    setReplyTo = false,
    None,
    configuration
  )
}

File: src/main/java/ru/tinkoff/gatling/javaapi/request/RequestReplyDslBuilderExchange.java
package ru.tinkoff.gatling.javaapi.request;

import static io.gatling.javaapi.core.internal.Expressions.*;

public class RequestReplyDslBuilderExchange {
    private final ru.tinkoff.gatling.amqp.request.RequestReplyDslBuilderExchange wrapped;

    public RequestReplyDslBuilderExchange(ru.tinkoff.gatling.amqp.request.RequestReplyDslBuilderExchange wrapped) {
        this.wrapped = wrapped;
    }

    public RequestReplyDslBuilderMessage directExchange(String name, String routingKey) {
        return new RequestReplyDslBuilderMessage(wrapped.directExchange(toStringExpression(name), toStringExpression(routingKey)));
    }

    public RequestReplyDslBuilderMessage topicExchange(String name, String routingKey) {
        return new RequestReplyDslBuilderMessage(wrapped.topicExchange(toStringExpression(name), toStringExpression(routingKey)));
    }

    public RequestReplyDslBuilderMessage queueExchange(String name) {
        return new RequestReplyDslBuilderMessage(wrapped.queueExchange(toStringExpression(name)));
    }

}
